pages/blog/typescript-types.mdx

```diff
@@ -0,0 +1,159 @@
+}
+export default withArticle(meta)
+ 
+ 
+# Introduction
+
+TypeScript programming language is a superset of JavaScript that adds types to JavaScript using a set of tools called a type system.
+All types inference assumes the expected type of the variable on code time. Type inference will log a complaint if the variable is later reassigned to a different type.
+For Example:
+
+```ts
+let pizzaName = 'Napolitana'
+
+pizzaName = 123 // Error: Type number (123) is not assignable to type 'string'.
+```
+
+When a variable is declared without being assigned an initial value, the default type is `any`.
+
+```ts
+let pizzaName
+
+pizzaName = 123 // Any got all types of values
+pizzaName = 'Napolitana' // Any got all types of values
+```
+
+Adding a type annpotation to a variable will force the variable to be of that type. It help you make sure the output will be the expected. And if not, you will get types error in code-time.
+
+```ts
+let pizzaName: string
+pizzaName = 'Napolitana'
+
+pizzaName = 123 // Error: Type number (123) is not assignable to type 'string'.
+```
+
+## Union Types
+
+After we understand how TypeScript type system works, we can start to understand how to use it to create a Union Type.
+TypeScript allow us a flexible type that called `any` type like we did in the previous example. But also TypeScript can combine multiple types and combine them together as a union type.
+
+```ts
+let pizzaLover: string | boolean
+
+pizzaLover = 'YES' // string type
+pizzaLover = true // boolean type
+```
+
+More Union type example:
+
+```ts
+type Sushi = {
+  kind: 'Sushi'
+  fish: string
+  amount: number
+}
+
+type Pizza = {
+  kind: 'Pizza'
+  topping: string
+  amount: number
+}
+
+type Food = Sushi | Pizza
+```
+
+So what is Type Narrowing?
+Type Narrowing is a process of narrowing a type to a more specific type.
+We can declare interface with all the properties of a type and then we can use that interface to narrow the type of a variable.
+Lets make basic narrowing example with function:
+
+```ts
+interface Food {
+  name: string
+  nutritionalValues?: {
+    calories: number
+    fat: number
+    protein: number
+    carbs: number
+  }
+}
+
+function getTotalCalories(f: Food): number | undefined {
+  if (f.nutritionalValues) {
+    f.nutritionalValues
+    // f.nutritionalValues is type of { calories: number, fat: number, protein: number, carbs: number }
+    return f.nutritionalValues.calories
+  }
+  f.nutritionalValues
+
+  return f.nutritionalValues
+}
+```
+
+### Operator `typeof`
+
+Varibale of a union type can assume one or even few diffrent types. In the example below we use `if` statement to check if the variable is of type `Food` or not.
+We can also use the opretor `typeof` with the variable name and compare it with the type you expect for the variable.
+Lets see a basic example of `typeof` operator in function:
+
+```ts
+function getScoreofLovingPizza(score: string | number) {
```

Comment:
I think maybe we can try to provide a more meaningful example here. 
Something like `number | "very-much"` and then translate the "very-much" string into `10` as number?